Installation issue: MPICH undefined reference

### Steps to reproduce the issue

```console
$ spack spec -l mpich +rocm amdgpu_target=gfx90a  ^rocm-core@6.0.0

Input spec
--------------------------------
 -   mpich+rocm amdgpu_target=gfx90a
 -       ^rocm-core@6.0.0

Concretized
--------------------------------
 -   ejrhcju  mpich@4.2.1%gcc@11.4.0~argobots~cuda+fortran+hwloc+hydra+libxml2+pci+rocm+romio~slurm~vci~verbs+wrapperrpath~xpmem amdgpu_target=gfx90a build_system=autotools datatype-engine=dataloop device=ch4 netmod=ofi pmi=default arch=linux-ubuntu22.04-icelake
[+]  cyqlta2      ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu22.04-icelake
[+]  j7q4m67      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-icelake
[e]  mwiwvx2      ^glibc@2.35%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  v4cjcg4      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-icelake
[+]  kob6g4o      ^hip@6.0.0%gcc@11.4.0~cuda~ipo+rocm build_system=cmake build_type=Release generator=make patches=1f65dfe,5bb9b0e,7668b2a,9165154,aee7249 arch=linux-ubuntu22.04-icelake
[+]  sdtqzzg          ^cmake@3.27.9%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-icelake
[+]  obpwmxc              ^curl@8.7.1%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-icelake
[+]  lmtdrlq                  ^nghttp2@1.57.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  wlzswt2          ^comgr@6.0.0%gcc@11.4.0~asan~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-icelake
[+]  djwgdfh              ^rocm-cmake@6.0.2%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-icelake
[+]  7wvogam          ^glx@1.4%gcc@11.4.0 build_system=bundle arch=linux-ubuntu22.04-icelake
[+]  4zxcivh              ^mesa@23.2.1%gcc@11.4.0+glx+llvm+opengl~opengles+osmesa~strip build_system=meson buildtype=release default_library=shared arch=linux-ubuntu22.04-icelake
[+]  2g3t6qi                  ^bison@3.8.2%gcc@11.4.0~color build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  dx6zwq2                  ^flex@2.6.3%gcc@11.4.0+lex~nls build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  t7bl4so                  ^glproto@1.4.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  neijvtd                  ^libunwind@1.6.2%gcc@11.4.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs~pic+tests+weak_backtrace~xz~zlib build_system=autotools components=none libs=shared,static arch=linux-ubuntu22.04-icelake
[+]  t42ngy4                  ^libx11@1.8.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  jnza3mz                      ^inputproto@2.3.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  yqjv36k                      ^kbproto@1.0.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  ylyyfsx                      ^xextproto@7.3.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  zjry5wh                      ^xproto@7.0.31%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  52bfv3p                      ^xtrans@1.5.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  5fvmmjp                  ^libxcb@1.16%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  t6h3npf                      ^libxau@1.0.11%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  4rgyg72                      ^libxdmcp@1.1.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  5hezlke                      ^xcb-proto@1.16.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  nh2o4a5                  ^libxext@1.3.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  tsasusb                  ^libxt@1.3.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  m63kylz                      ^libice@1.1.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  irvgvsq                      ^libsm@1.2.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  vaqanp2                  ^llvm@17.0.6%gcc@11.4.0+clang~cuda~flang+gold~ipo+libomptarget~libomptarget_debug~link_llvm_dylib+lld+lldb+llvm_dylib+lua~mlir+polly~python~split_dwarf~z3~zstd build_system=cmake build_type=Release compiler-rt=runtime generator=ninja libcxx=runtime libunwind=runtime openmp=runtime shlib_symbol_version=none targets=all version_suffix=none arch=linux-ubuntu22.04-icelake
[+]  hdgkqmg                      ^binutils@2.42%gcc@11.4.0~gas+gold~gprofng+headers~interwork+ld~libiberty~lto~nls~pgo+plugins build_system=autotools compress_debug_sections=zlib libs=shared,static arch=linux-ubuntu22.04-icelake
[+]  bibdyqc                      ^libedit@3.1-20230828%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  cxtt6cs                      ^lua@5.3.6%gcc@11.4.0~pcfile+shared build_system=makefile fetcher=curl arch=linux-ubuntu22.04-icelake
[+]  ftce7wh                          ^unzip@6.0%gcc@11.4.0 build_system=makefile patches=f6f6236 arch=linux-ubuntu22.04-icelake
[+]  zta6fzi                      ^perl-data-dumper@2.173%gcc@11.4.0 build_system=perl arch=linux-ubuntu22.04-icelake
[+]  vtnviy7                      ^swig@4.1.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  qsudn5t                          ^pcre2@10.42%gcc@11.4.0~jit+multibyte build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  ogx6yxx                  ^meson@1.3.2%gcc@11.4.0 build_system=python_pip patches=0f0b1bd arch=linux-ubuntu22.04-icelake
[+]  icqi6gq                  ^py-mako@1.2.4%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-icelake
[+]  4eo6iyu                      ^py-markupsafe@2.1.3%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-icelake
[+]  fr24lea                  ^xrandr@1.5.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  kxhlyrr                      ^libxrandr@1.5.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  gmffi6p                          ^renderproto@0.11.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  hb44tbl                      ^libxrender@0.9.11%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  ih4524n                      ^randrproto@1.5.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  qroen43          ^hipcc@6.0.2%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make patches=f497181 arch=linux-ubuntu22.04-icelake
[+]  l2yp4v3          ^hipify-clang@6.0.0%gcc@11.4.0~asan~ipo build_system=cmake build_type=Release generator=make patches=54b8b39 arch=linux-ubuntu22.04-icelake
[+]  n3bhye3          ^hsakmt-roct@6.0.0%gcc@11.4.0~asan~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-icelake
[+]  gkl75ec              ^libdrm@2.4.120%gcc@11.4.0~docs build_system=generic arch=linux-ubuntu22.04-icelake
[+]  g3nk2f2                  ^libpthread-stubs@0.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  mbkizob          ^numactl@2.0.14%gcc@11.4.0 build_system=autotools patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu22.04-icelake
[+]  eqea3wx              ^autoconf@2.72%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  j2fwy5g              ^automake@1.16.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  qoolmq5              ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-icelake
[+]  hhdqrnz                  ^diffutils@3.10%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  xr76aco                  ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  6evzo2p          ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-icelake
[+]  ntfaayd              ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-icelake
[+]  3qvfkw2              ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-icelake
[+]  yythx2o              ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  byod3ze          ^perl-file-which@1.27%gcc@11.4.0 build_system=perl arch=linux-ubuntu22.04-icelake
[+]  ai2smpw          ^perl-uri-encode@1.1.1%gcc@11.4.0 build_system=perl arch=linux-ubuntu22.04-icelake
[+]  hbvluen              ^perl-module-build@0.4232%gcc@11.4.0 build_system=perl arch=linux-ubuntu22.04-icelake
[+]  ghgf4xk          ^py-cppheaderparser@2.7.4%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-icelake
[+]  bz6567b              ^py-pip@23.1.2%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-icelake
[+]  w7ijpwd              ^py-ply@3.11%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-icelake
[+]  vgxuj5c              ^py-setuptools@69.2.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-icelake
[+]  x4vv627              ^py-wheel@0.41.2%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-icelake
[+]  zm4adrz          ^rocm-core@6.0.0%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-icelake
[+]  c6u3foe          ^rocminfo@6.0.0%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-icelake
[+]  yt2owq7          ^roctracer-dev-api@6.0.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-icelake
[+]  vjwaeny      ^hsa-rocr-dev@6.0.0%gcc@11.4.0~asan+image~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-icelake
[+]  tgabm4h          ^elfutils@0.190%gcc@11.4.0~debuginfod+exeprefix+nls build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  a4hu642              ^gettext@0.22.4%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  6jxvrsx                  ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu22.04-icelake
[+]  ygc6pq2                      ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-icelake
[+]  xctpcie              ^zstd@1.5.6%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-icelake
[+]  db2pwyr          ^xxd-standalone@8.2.1201%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-icelake
[+]  c23r2o2      ^hwloc@2.9.1%gcc@11.4.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu22.04-icelake
[+]  ub2updv          ^ncurses@6.4%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu22.04-icelake
[+]  76i2mdj      ^libfabric@1.21.0%gcc@11.4.0~debug~kdreg~uring build_system=autotools fabrics=sockets,tcp,udp arch=linux-ubuntu22.04-icelake
[+]  msq4y3l      ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  bjuewgv          ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  fmv325e          ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  a6bxg2s      ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  kkekvvl          ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-icelake
[+]  mevq4mv          ^zlib-ng@2.1.6%gcc@11.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  uexqpap      ^llvm-amdgpu@6.0.0%gcc@11.4.0~ipo~link_llvm_dylib~llvm_dylib+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=53f9500,b4774ca arch=linux-ubuntu22.04-icelake
[+]  pktg7r2          ^ninja@1.11.1%gcc@11.4.0+re2c build_system=generic arch=linux-ubuntu22.04-icelake
[+]  ipuehfp              ^re2c@2.2%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-icelake
[+]  zwaaixh          ^python@3.11.7%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu22.04-icelake
[+]  6fl2cmq              ^expat@2.6.2%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  zhtwnsz                  ^libbsd@0.12.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  k6vfcho                      ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  qbozdnm              ^libffi@3.4.6%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  wh65ywm              ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu22.04-icelake
[+]  5lzsyst              ^openssl@3.2.1%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-icelake
[+]  yvu7abt                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-icelake
[+]  o6tb2ki              ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-icelake
[+]  qudorvw              ^sqlite@3.43.2%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  yzjfdvu              ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake
[+]  4jpxjdg          ^z3@4.12.4%gcc@11.4.0~gmp~ipo~python build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-icelake
[+]  vmnvn4t      ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-icelake


```


### Error message

<details><summary>Error message</summary>

<pre>
Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

8 errors found in build log:
     12216    /bin/bash ./libtool  --tag=CXX   --mode=link /storage/users/jalcaraz/spackkkk/spack/lib/spack/env/gcc/g++  -O2 -version-info 16:1:4 -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/libfabric-1.21.0-
              76i2mdj23tspkxvws3p6iovjumbpdotz/lib -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/hwloc-2.9.1-c23r2o2keiuyfzehu47zz3bvia33x75e/lib -o lib/libmpicxx.la -rpath /storage/users/jalcaraz/spackkkk/spa
              ck/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/mpich-4.2.1-ejrhcjukenc4l6melae3q4n6pmwawlsi/lib src/binding/cxx/initcxx.lo lib/libmpi.la
     12217    /bin/bash ./libtool  --tag=CC   --mode=link /storage/users/jalcaraz/spackkkk/spack/lib/spack/env/gcc/gcc -fvisibility=hidden -O2  -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/libfabric-1.21.0-76
              i2mdj23tspkxvws3p6iovjumbpdotz/lib -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/hwloc-2.9.1-c23r2o2keiuyfzehu47zz3bvia33x75e/lib -o src/env/mpichversion src/env/mpichversion.o lib/libmpi.la
     12218    /bin/bash ./libtool  --tag=CC   --mode=link /storage/users/jalcaraz/spackkkk/spack/lib/spack/env/gcc/gcc -fvisibility=hidden -O2  -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/libfabric-1.21.0-76
              i2mdj23tspkxvws3p6iovjumbpdotz/lib -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/hwloc-2.9.1-c23r2o2keiuyfzehu47zz3bvia33x75e/lib -o src/env/mpivars src/env/mpivars.o lib/libmpi.la
     12219    libtool: link: /storage/users/jalcaraz/spackkkk/spack/lib/spack/env/gcc/gcc -fvisibility=hidden -O2 -o src/env/.libs/mpivars src/env/mpivars.o  -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/libfa
              bric-1.21.0-76i2mdj23tspkxvws3p6iovjumbpdotz/lib -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/hwloc-2.9.1-c23r2o2keiuyfzehu47zz3bvia33x75e/lib lib/.libs/libmpi.so -lhwloc -lfabric -latomic -ldl
              -lpthread -lze_loader -Wl,-rpath -Wl,/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/mpich-4.2.1-ejrhcjukenc4l6melae3q4n6pmwawlsi/lib
     12220    libtool: link: /storage/users/jalcaraz/spackkkk/spack/lib/spack/env/gcc/gcc -fvisibility=hidden -O2 -o src/env/.libs/mpichversion src/env/mpichversion.o  -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11
              .4.0/libfabric-1.21.0-76i2mdj23tspkxvws3p6iovjumbpdotz/lib -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/hwloc-2.9.1-c23r2o2keiuyfzehu47zz3bvia33x75e/lib lib/.libs/libmpi.so -lhwloc -lfabric -lat
              omic -ldl -lpthread -lze_loader -Wl,-rpath -Wl,/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/mpich-4.2.1-ejrhcjukenc4l6melae3q4n6pmwawlsi/lib
     12221    libtool: link: /storage/users/jalcaraz/spackkkk/spack/lib/spack/env/gcc/g++  -fPIC -DPIC -shared -nostdlib /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/11/crtbeginS.o  src/binding/cxx/.l
              ibs/initcxx.o   -Wl,-rpath -Wl,/tmp/jalcaraz/spack-stage/spack-stage-mpich-4.2.1-ejrhcjukenc4l6melae3q4n6pmwawlsi/spack-src/lib/.libs -Wl,-rpath -Wl,/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/mp
              ich-4.2.1-ejrhcjukenc4l6melae3q4n6pmwawlsi/lib -L/storage/users/jalcaraz/spackkkk/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/libfabric-1.21.0-76i2mdj23tspkxvws3p6iovjumbpdotz/lib -L/storage/users/jalcaraz/spackkkk/spack/opt/spa
              ck/linux-ubuntu22.04-icelake/gcc-11.4.0/hwloc-2.9.1-c23r2o2keiuyfzehu47zz3bvia33x75e/lib lib/.libs/libmpi.so -lhwloc -lfabric -latomic -ldl -lpthread -lze_loader -L/usr/lib/gcc/x86_64-linux-gnu/11 -L/usr/lib/gcc/x86_64-linux-gnu/11/../..
              /../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/11/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/11/../../.. -lstdc++ -lm -lc -lgcc_s /usr/lib/gcc
              /x86_64-linux-gnu/11/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crtn.o  -O2   -Wl,-soname -Wl,libmpicxx.so.12 -o lib/.libs/libmpicxx.so.12.4.1
  >> 12222    /usr/bin/ld: lib/.libs/libmpi.so: undefined reference to `MPIR_Typerep_test'
  >> 12223    collect2: error: ld returned 1 exit status
  >> 12224    make[2]: *** [Makefile:10166: src/env/mpichversion] Error 1
     12225    make[2]: *** Waiting for unfinished jobs....
  >> 12226    /usr/bin/ld: lib/.libs/libmpi.so: undefined reference to `MPIR_Typerep_test'
  >> 12227    collect2: error: ld returned 1 exit status
  >> 12228    make[2]: *** [Makefile:10172: src/env/mpivars] Error 1
     12229    libtool: link: (cd "lib/.libs" && rm -f "libmpicxx.so.12" && ln -s "libmpicxx.so.12.4.1" "libmpicxx.so.12")
     12230    libtool: link: (cd "lib/.libs" && rm -f "libmpicxx.so" && ln -s "libmpicxx.so.12.4.1" "libmpicxx.so")
     12231    libtool: link: ar cru lib/.libs/libmpicxx.a  src/binding/cxx/initcxx.o
     12232    ar: `u' modifier ignored since `D' is the default (see `U')
     12233    libtool: link: ranlib lib/.libs/libmpicxx.a
     12234    libtool: link: ( cd "lib/.libs" && rm -f "libmpicxx.la" && ln -s "../libmpicxx.la" "libmpicxx.la" )
     12235    make[2]: Leaving directory '/tmp/jalcaraz/spack-stage/spack-stage-mpich-4.2.1-ejrhcjukenc4l6melae3q4n6pmwawlsi/spack-src'
  >> 12236    make[1]: *** [Makefile:19962: all-recursive] Error 1
     12237    make[1]: Leaving directory '/tmp/jalcaraz/spack-stage/spack-stage-mpich-4.2.1-ejrhcjukenc4l6melae3q4n6pmwawlsi/spack-src'
  >> 12238    make: *** [Makefile:5815: all] Error 2

</pre></details>


### Information on your system

* **Spack:** 0.22.0.dev0 (de5b4840e96bb026bf9c0a2f0d4114b15c0697df)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-icelake
* **Concretizer:** clingo


### Additional information

@raffenet  @yfguo

[spack-build-env.txt](https://github.com/spack/spack/files/15259534/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/15259535/spack-build-out.txt)


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/mpich/package.py
+++ b/var/spack/repos/builtin/packages/mpich/package.py
@@ -135,6 +135,8 @@ supported, and netmod is ignored if device is ch3:sock.""",
 
     conflicts("datatype-engine=yaksa", when="device=ch3")
     conflicts("datatype-engine=yaksa", when="device=ch3:sock")
+    conflicts("datatype-engine=dataloop", when="+cuda")
+    conflicts("datatype-engine=dataloop", when="+rocm")
 
     variant(
         "hcoll",

